subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/NormalizedPathFingerprintCompareStrategy.java

```diff
@@ -109,10 +109,13 @@ private boolean doVisitChangesSince(
         return true;
     }
 
-    // There might be multiple files with the same normalized path, here we choose one of them
+    // There might be multiple files with the same normalized path, here we choose one of them (favoring absolute path matches)
     private static boolean wasModified(ListMultimap<String, FilePathWithType> addedFilesByNormalizedPath, String normalizedPath, FilePathWithType pathWithType) {
-        List<FilePathWithType> addedFilesForNormalizedPath = addedFilesByNormalizedPath.get(normalizedPath);
-        return !addedFilesForNormalizedPath.isEmpty() && addedFilesForNormalizedPath.remove(0) != null;
+        return addedFilesByNormalizedPath.get(normalizedPath)
+            .removeIf(file ->
+                file.getFileType() != FileType.Missing
```

Comment:
Otherwise deleted files are marked as modified (only their type changes on deletion).
Previously the `isEmpty` check took care of that.

Reply:
I could add the ismempty back, if you think it's better - asked @wolfs explicitly, he said it's more descriptive this way

Reply:
Update: reverting all changes and running the new integration test that removes files, it seems this behavior wasn't correct before: removed files were calling the `DefaultFileChange.modified` method with `FileType.Missing` and `ChangeTypeInternal.MODIFIED`, instead of `REMOVED`:
<img width="1364" alt="Screenshot 2019-05-20 at 23 17 42" src="https://user-images.githubusercontent.com/1841944/58052595-8e384600-7b55-11e9-9688-f1e19087475e.png">

The change you've mentioned was needed for it to be detected correctly (i.e. the refactoring revealed the bug):
<img width="641" alt="Screenshot 2019-05-20 at 23 24 15" src="https://user-images.githubusercontent.com/1841944/58052954-7614f680-7b56-11e9-81f6-c64341ae63a7.png">

Reply:
Update2: @wolfs said this is actually a feature, so I've reorganized the code to reflect that.
Thanks for your inputs.